src/cpp/src/visual_language/input_embedder.cpp

```diff
@@ -860,6 +860,108 @@ class InputsEmbedderLLaVANext : public InputsEmbedderLLaVA {
+            std::string image_context_token = m_vlm_config.image_context_token;
+            std::string image_end_token = m_vlm_config.image_end_token;
+
+            const size_t num_patches = image_embeds.get_shape().at(0);
+            const size_t num_image_tokens = image_embeds.get_shape().at(1);
+
+            std::string formatted_prompt = image_token + "\n" + prompt;
+            
+            std::string concated_image_tokens;
+            concated_image_tokens += image_start_token;
+            for (int i = 0; i < num_patches * num_image_tokens; ++i) {
+                concated_image_tokens += image_context_token;
+            }
+            concated_image_tokens += image_end_token;
+
+            size_t pos = formatted_prompt.find(image_token);
+            if (pos != std::string::npos) {
+                formatted_prompt.replace(pos, image_token.length(), concated_image_tokens);
+            }
+            
+            ov::Tensor input_ids = get_encoded_input_ids(formatted_prompt);
+            ov::Tensor text_embeds = process_prompt(m_embedding, input_ids, m_vlm_config.scale_emb);
+
+            ov::Tensor encoded_image_context_token = m_tokenizer.encode(image_context_token, ov::genai::add_special_tokens(false)).input_ids;
+            int64_t image_context_token_id = encoded_image_context_token.data<int64_t>()[encoded_image_context_token.get_size() - 1];
+
+            return merge_text_and_image_embeddings_internvl(input_ids, text_embeds, image_embeds, image_context_token_id);
+        }
+    }
+
+protected:
+    ov::Tensor merge_text_and_image_embeddings_internvl(
+        const ov::Tensor& input_ids,
+        const ov::Tensor& text_embeds,
+        const ov::Tensor& image_embeds,
+        int64_t image_context_token_id
+    ) {
+        auto text_embeds_shape = text_embeds.get_shape();
+        auto image_embeds_shape = image_embeds.get_shape();
+        size_t batch_size = text_embeds_shape.at(0);
+        size_t seq_len = text_embeds_shape.at(1);
+        size_t embed_dim = text_embeds_shape.at(2);
+
+        ov::Tensor merged_embeds(text_embeds.get_element_type(), text_embeds_shape);
+
+        const float* image_embeds_data = image_embeds.data<float>();
+        const float* text_embeds_data = text_embeds.data<float>();
+        const int64_t* input_ids_data = input_ids.data<int64_t>();
+        float* merged_embeds_data = merged_embeds.data<float>();
+
+        size_t flattened_size = batch_size * seq_len;
+        std::vector<bool> image_context_tokens_mask(flattened_size, false);
+        size_t image_context_tokens_count = 0;
+
+        for (size_t i = 0; i < flattened_size; ++i) {
+            if (input_ids_data[i] == image_context_token_id) {
+                image_context_tokens_mask[i] = true;
+                ++image_context_tokens_count;
+            }
+        }
+
+        OPENVINO_ASSERT(image_context_tokens_count > 0, "input_ids does not contain image context token ids");
```

Comment:
does it mean that we cannot perform generate w/o image? as `image_context_tokens_count` is required

Reply:
This part of code belongs to `merge_text_and_image_embeddings_internvl` method, that is executed only when images are provided and not empty.
So generate call w/o an image works